File: lib/ts/src/parent.ts
import * as net from 'node:net';
import * as os from 'node:os';
import * as path from 'node:path';
import * as fs from 'node:fs';
import * as crypto from 'node:crypto';
import {type ChildProcess, spawn} from 'node:child_process';
import {IPCCommon, type IPCOptions} from './common.js';
import {timeout} from './util.js';

const IPC_SOCKET_ARG = 'ipc-socket';
const ACCEPT_TIMEOUT_MS = 10000;

export class ParentIPC extends IPCCommon {
    private readonly cmdPath: string;
    private readonly cmdArgs: string[];
    private cmd: ChildProcess | null = null;
    private readonly listener: net.Server;
    private cmdExitResult: { code: number | null, signal: string | null } | null = null;
    private cmdExitCallbacks: ((result: { code: number | null, signal: string | null }) => void)[] = [];

    constructor(cmdPath: string, cmdArgs: string[], opts?: IPCOptions, ...localApis: object[]) {
        const socketPath = path.join(os.tmpdir(), `kitten-ipc-${ process.pid }-${ crypto.randomInt(2**48 - 1) }.sock`);
        super(localApis, socketPath, opts);

        this.cmdPath = cmdPath;
        if (cmdArgs.includes(`--${ IPC_SOCKET_ARG }`)) {
            throw new Error(`you should not use '--${ IPC_SOCKET_ARG }' argument in your command`);
        }
        this.cmdArgs = cmdArgs;

        this.listener = net.createServer();
    }

    async start(): Promise<void> {
        try {
            fs.unlinkSync(this.socketPath);
        } catch {
        }

        await new Promise<void>((resolve, reject) => {
            this.listener.listen(this.socketPath, () => {
                resolve();
            });
            this.listener.on('error', reject);
        });

        const cmdArgs = [...this.cmdArgs, `--${ IPC_SOCKET_ARG }`, this.socketPath];
        this.cmd = spawn(this.cmdPath, cmdArgs, {stdio: 'inherit'});

        this.cmd.on('error', (err) => {
            this.raiseErr(err);
        });

        this.cmd.on('close', (code, signal) => {
            const result = { code, signal };
            this.cmdExitResult = result;
            for (const cb of this.cmdExitCallbacks) cb(result);
            this.cmdExitCallbacks = [];
        });

        await this.acceptConn();
    }

    private async acceptConn(): Promise<void> {
        const acceptPromise = new Promise<net.Socket>((resolve, reject) => {
            this.listener.once('connection', (conn) => {
                resolve(conn);
            });
            this.listener.once('error', reject);
        });

        const exitPromise = new Promise<net.Socket>((_, reject) => {
            if (this.cmdExitResult) {
                reject(new Error(`command exited before connection established`));
            } else {
                this.cmdExitCallbacks.push(() => {
                    reject(new Error(`command exited before connection established`));
                });
            }
        });

        try {
            this.conn = await timeout(Promise.race([acceptPromise, exitPromise]), ACCEPT_TIMEOUT_MS);
            this.readConn();
        } catch (e) {
            if (this.cmd) this.cmd.kill();
            throw e;
        }
    }

    async wait(): Promise<void> {
        if (!this.cmd) {
            throw new Error('Command is not started yet');
        }

        const exitPromise = new Promise<{ code: number | null, signal: string | null }>((resolve) => {
            if (this.cmdExitResult) {
                resolve(this.cmdExitResult);
            } else {
                this.cmdExitCallbacks.push(resolve);
            }
        });

        try {
            await Promise.race([
                exitPromise.then(({ code, signal }) => {
                    if (signal || code) {
                        if (signal) throw new Error(`Process exited with signal ${ signal }`);
                        else throw new Error(`Process exited with code ${ code }`);
                    } else if (!this.ready) {
                        throw new Error('command exited before connection established');
                    }
                }),
                this.errorQueue.collect().then((errors) => {
                    if (errors.length === 1) {
                        throw errors[0];
                    } else if (errors.length > 1) {
                        throw new Error(errors.map(e => e.toString()).join(', '));
                    }
                }),
            ]);
        } finally {
            try { fs.unlinkSync(this.socketPath); } catch {}
        }
    }
}

File: lib/golang/parent.go
package golang

import (
	"context"
	"errors"
	"fmt"
	"math/rand"
	"net"
	"os"
	"os/exec"
	"path/filepath"
	"slices"
	"syscall"
	"time"
)

type ParentIPC struct {
	*ipcCommon
	cmd      *exec.Cmd
	listener net.Listener
	cmdDone  chan struct{}
	cmdErr   error
}

func NewParent(cmd *exec.Cmd, localApis ...any) (*ParentIPC, error) {
	return NewParentWithContext(context.Background(), cmd, localApis...)
}

func NewParentWithContext(ctx context.Context, cmd *exec.Cmd, localApis ...any) (*ParentIPC, error) {
	p := ParentIPC{
		ipcCommon: &ipcCommon{
			localApis:    mapTypeNames(localApis),
			pendingCalls: make(map[int64]*pendingCall),
			errCh:        make(chan error, 1),
			socketPath:   filepath.Join(os.TempDir(), fmt.Sprintf("kitten-ipc-%d-%d.sock", os.Getpid(), rand.Int63())),
			ctx:          ctx,
		},
		cmd: cmd,
	}

	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr

	if slices.Contains(cmd.Args, ipcSocketArg) {
		return nil, fmt.Errorf("you should not use `%s` argument in your command", ipcSocketArg)
	}
	cmd.Args = append(cmd.Args, ipcSocketArg, p.socketPath)

	p.errCh = make(chan error, 1)
	p.cmdDone = make(chan struct{})

	return &p, nil
}

func (p *ParentIPC) Start() error {
	_ = os.Remove(p.socketPath)
	listener, err := net.Listen("unix", p.socketPath)
	if err != nil {
		return fmt.Errorf("listen unix socket: %w", err)
	}
	p.listener = listener
	defer p.listener.Close()

	err = p.cmd.Start()
	if err != nil {
		return fmt.Errorf("cmd start: %w", err)
	}

	go func() {
		p.cmdErr = p.cmd.Wait()
		close(p.cmdDone)
	}()

	return p.acceptConn()
}

type connResult struct {
	conn net.Conn
	err  error
}

func (p *ParentIPC) acceptConn() error {
	res := make(chan connResult, 1)
	go func() {
		conn, err := p.listener.Accept()
		res <- connResult{conn: conn, err: err}
		close(res)
	}()

	select {
	case <-time.After(time.Duration(defaultAcceptTimeout) * time.Second):
		_ = p.cmd.Process.Kill()
		return fmt.Errorf("accept timeout")
	case <-p.cmdDone:
		return fmt.Errorf("cmd exited before accepting connection: %w", p.cmdErr)
	case r := <-res:
		if r.err != nil {
			_ = p.cmd.Process.Kill()
			return fmt.Errorf("accept: %w", r.err)
		}
		p.conn = r.conn
		go p.readConn()
	}
	return nil
}

func (p *ParentIPC) Stop() error {
	p.mu.Lock()
	hasPending := len(p.pendingCalls) > 0
	p.mu.Unlock()
	if hasPending {
		return fmt.Errorf("there are calls pending")
	}
	if p.processingIncomingCalls.Load() > 0 {
		return fmt.Errorf("there are calls processing")
	}
	p.stopRequested.Store(true)
	if err := p.cmd.Process.Signal(syscall.SIGINT); err != nil {
		return fmt.Errorf("send SIGINT: %w", err)
	}
	return p.Wait()
}

func (p *ParentIPC) Wait(timeout ...time.Duration) (retErr error) {
	const maxDuration = time.Duration(1<<63 - 1)
	_timeout := maxDuration
	if len(timeout) > 0 {
		_timeout = timeout[0]
	}

loop:
	for {
		select {
		case err := <-p.errCh:
			retErr = mergeErr(retErr, fmt.Errorf("ipc internal error: %w", err))
			break loop
		case <-p.cmdDone:
			err := p.cmdErr
			if err != nil {
				var exitErr *exec.ExitError
				if ok := errors.As(err, &exitErr); ok {
					if !exitErr.Success() {
						ws, ok := exitErr.Sys().(syscall.WaitStatus)
						if !(ok && ws.Signaled() && ws.Signal() == syscall.SIGINT && p.stopRequested.Load()) {
							retErr = mergeErr(retErr, fmt.Errorf("cmd wait: %w", err))
						}
					}
				} else {
					retErr = mergeErr(retErr, fmt.Errorf("cmd wait: %w", err))
				}
			}
			break loop
		case <-time.After(_timeout):
			p.stopRequested.Store(true)
			if err := p.cmd.Process.Signal(syscall.SIGINT); err != nil {
				retErr = mergeErr(retErr, fmt.Errorf("send SIGINT: %w", err))
			}
		}
	}

	p.closeConn()
	_ = os.Remove(p.socketPath)

	return retErr
}
